src/mgr/Mgr.cc

```diff
@@ -57,17 +57,27 @@ Mgr::Mgr(MonClient *monc_, const MgrMap& mgrmap,
   objecter(objecter_),
   client(client_),
   client_messenger(clientm_),
-  finisher(g_ceph_context, "Mgr", "mgr-fin"),
+  finishers{{"generic", std::make_shared<Finisher>(g_ceph_context, "Mgr", "mgr-fin-generic")}},
   digest_received(false),
   py_module_registry(py_module_registry_),
   cluster_state(monc, nullptr, mgrmap),
-  server(monc, finisher, daemon_state, cluster_state, *py_module_registry,
+  server(monc, finishers, daemon_state, cluster_state, *py_module_registry,
          clog_, audit_clog_),
   clog(clog_),
   audit_clog(audit_clog_),
   initialized(false),
   initializing(false)
 {
+  auto module_names = py_module_registry->get_module_names();
+  std::string thread_name;
+  for (const auto& module_name : module_names) {
+    dout(1) << "Initializing module finisher thread '" << module_name << "'" << dendl;
+    thread_name = "fin-" + module_name;
+    if (thread_name.size() > 15) {
+        thread_name.resize(15);
+    }
+    finishers.emplace(module_name, std::make_shared<Finisher>(g_ceph_context, "Mgr", thread_name));
```

Comment:
Why not make the finisher  part of the module class?

Reply:
Do you mean part of `PyModuleRegistry` class ? The `finishers` is a single map which contains pointers to generic and per module finishers. The reference to this finishers map is required across different classes like `DeamonSever`, `ActivePyModules`, `PyModuleRegistry`.  The `Mgr::init()` (active_mgr) is called via `MgrStandby::handle_mgr_map()` via active mgr's finisher thread. How does it help if finisher is part of `PyModuleRegistry` class ?

Reply:
No, part of `ActivePyModule` (I think that's the best place). Then you can lookup the module and queue it there.

Reply:
@batrick Done. PTAL.